### PR TITLE
Allow non-interpolatory elements in FESystem for generalized supp.p.

### LINKED ISSUE #5438: VectorTools::interpolate still needs fixing for systems where parts have no support points
#5334 introduces a fallback implementation for the case of system elements where parts have no support points (think Q2 x DGP1 for Stokes). This was necessary because #4857 broke this functionality. 

#5334 is only intended as a stop-gap. The correct way is to implement the necessary functionality in the new implementation introduced in #4857, not to duplicate a few hundred lines of code. @tamiko promised to fix this issue in #5334.

### PR DESCRIPTION
**Pull request by tamiko**
Closes #5438
In reference to #5334
In reference to #4857

### CODE REVIEW

File: source/fe/fe_system.cc [hunk 1]
```diff
@@ -2302,6 +2303,29 @@ convert_generalized_support_point_values_to_dof_values (const std::vector<Vector
...
+
+          // To do this, we could really use a base_to_system_index()
+          // function, but that doesn't exist -- just do it by using the
+          // reverse table -- the amount of work done here is not worth
+          // trying to optimizing this.
+          for (unsigned int m = 0; m < multiplicity; ++m)
+            for (unsigned int i = 0; i < this->dofs_per_cell; ++i)
+              if (this->system_to_base_index(i).first ==
+                  std::make_pair(base, m))
+                dof_values[i] = std::numeric_limits<double>::quiet_NaN();
```
Comment (tamiko) [inline source/fe/fe_system.cc]: I am not sure whether it is safe to use a `signaling_NaN` here. Or whether we actually want that.

File: tests/fe/interpolate_q_system_mixed_01.debug.output [hunk 2]
```diff
@@ -8,7 +8,7 @@ DEAL::FESystem<1>[FE_Q<1>(1)^2-FE_Q<1>(2)-FE_DGP<1>(1)], Mask 2, P_2, rel. error
 DEAL::FESystem<1>[FE_Q<1>(1)^2-FE_Q<1>(2)-FE_DGP<1>(1)], Mask 1, P_3, rel. error=0.000296
 DEAL::FESystem<1>[FE_Q<1>(1)^2-FE_Q<1>(2)-FE_DGP<1>(1)], Mask 2, P_3, rel. error=0.000296
 DEAL::FESystem<1>[FE_Q<1>(1)^2-FE_Q<1>(2)-FE_DGP<1>(1)], Fails for including non-interpolating FE OK
-DEAL::	Fails with ExcNonInterpolatingFE()
+DEAL::	Fails with ExcMessage("The component mask supplied to " "VectorTools::interpolate selects a " "non-interpolatory element.")
```
Comment (tamiko) [inline tests/fe/interpolate_q_system_mixed_01.debug.output]: In spirit of our crusade specialized typed exceptions, I simply replaced `ExcNonInterpolatingFE` with an `ExcMessage`.

**Comment by tamiko** [comment]
@bangerth @gassmoeller *ping* I think this PR is a cleaner approach for supporting non-interpolatory FE in an `FESystem`. The idea is to simply fill the supplied vector of dof values with NaNs for non-interpolatory base elements. I have added a check to `VectorTools::interpolate` to raise an exception if such a non-interpolatory base element gets selected.

File: tests/fe/interpolate_q_system_mixed_01.debug.output (hunk 2 above)
Comment (bangerth) [reply tests/fe/interpolate_q_system_mixed_01.debug.output]: Appreciated! (In reference to #610 :-) )

File: source/fe/fe_system.cc (hunk 1 above)
Comment (bangerth) [reply source/fe/fe_system.cc]: I like using a `numbers::signaling_nan<double>`. -- in essence, you are saying "*** these values are known to be invalid ***" in the return object. That makes sense to me.

But this should be documented in the function's public interface. The only complication (also worth noting in the function doc) is that one cannot test for a value to be a signaling NaN. In general, one shouldn't even try to *read* it. But one doesn't have to -- you will know which elements of the output vector to ignore in the place where you're calling the current function

File: source/fe/fe_system.cc [hunk 3]
```diff
@@ -2302,6 +2303,29 @@ convert_generalized_support_point_values_to_dof_values (const std::vector<Vector
...
+
+          current_vector_component += multiplicity * n_base_components;
+          continue;
```
Comment (bangerth) [inline source/fe/fe_system.cc]: I'd rather you use `else { ... }` for the remaining part of this block.

Or, even better, reverse the two cases and treat the situation that we *have* generalized support points first in the `if`, and the less likely case in the `else`.

**Comment by bangerth** [comment]
Great, I like this solution. This is probably not only the easiest approach, but also the most elegant one. I would have tried this in the `interpolate()` function, but think that your solution is much better!

**Comment by tamiko** [comment]
@bangerth Updated.

**Comment by bangerth** [comment]
Thanks!

**Comment by gassmoeller** [comment]
Seems to work. Many thanks!

**Pull request merged**